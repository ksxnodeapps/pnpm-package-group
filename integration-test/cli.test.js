'use strict'
const path = require('path')
const fs = require('fs')
const process = require('process')
const {spawnSync} = require('child_process')
const fsForce = require('fs-force')
const {envMod} = require('../lib.dev/path-env')
const executable = require.resolve('../index.js')

const env = envMod()
  .surround(path.resolve(__dirname, 'virtual-env/bin'))
  .get()

const trackSpawnSnap = (argv = [], options = {}) => () => {
  const {status, error, signal, stdout, stderr} = spawnSync(executable, argv, {encoding: 'utf8', env, ...options})
  expect({status, error, signal, stdout, stderr}).toMatchSnapshot()
}

const realpath = name =>
  path.resolve(__dirname, name)

const filetext = name =>
  fs.readFileSync(realpath(name), 'utf8')

describe('program', () => {
  const workingDirectory = path.resolve(__dirname, 'virtual-env/working-directory')
  fsForce.deleteSync(workingDirectory)
  fsForce.mkdirSync(workingDirectory)
  process.chdir(workingDirectory)

  it('--help', trackSpawnSnap(['--help']))
  it('being invoked with neither arguments nor stdin ', trackSpawnSnap())

  describe('being invoked with stdin', () => {
    const mkopt = file => ({
      input: filetext(file)
    })

    const mkfn = file =>
      trackSpawnSnap([], mkopt(file))

    it('which contain valid syntax and schema', mkfn('input/valid.yaml'))
    it('which contain valid syntax but invalid schema', mkfn('input/invalid-schema.yaml'))
    it('which contain invalid syntax', mkfn('input/invalid-syntax.txt'))
  })

  describe('being invoked with paths', () => {
    const fn = (...xpath) => describe(xpath.join(' '), () => {
      it('without options', trackSpawnSnap(xpath))

      it(
        '--local=explicitly-specified-target',
        trackSpawnSnap(['--local=explicitly-specified-target', ...xpath])
      )
    })

    fn('../../input/valid.yaml')
    fn('../../input/invalid-schema.yaml')
    fn('../../input/invalid-syntax.txt')

    fn('../../input/valid.yaml', '../../input/invalid-schema.yaml')
    fn('../../input/valid.yaml', '../../input/invalid-syntax.txt')
    fn('../../input/invalid-schema.yaml', '../../input/invalid-syntax.txt')
    fn('../../input/valid.yaml', '../../input/invalid-schema.yaml', '../../input/invalid-syntax.txt')

    fn('../../input/valid.yaml/Nested')
    fn('../../input/valid.yaml/Flat')
    fn('../../input/valid.yaml/Global')
    fn('../../input/valid.yaml/DividedFlat')
    fn('../../input/valid.yaml/SelectiveNested')

    fn('../../input/valid.yaml/Nested/c')
    fn('../../input/valid.yaml/Flat/i/l')
    fn('../../input/valid.yaml/Global/i/l/p')
  })
})
