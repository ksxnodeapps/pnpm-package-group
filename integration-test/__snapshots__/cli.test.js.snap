// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`program --help 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "$ index.js [options] [arguments]

Options:
  --version            Show version number                             [boolean]
  --local, -l          Directory contains local packages  [string] [default: \\"\\"]
  --pnpm               Program to call                [string] [default: \\"pnpm\\"]
  --shell              Whether shell is used           [boolean] [default: true]
  --quiet-pnpm         Don't print pnpm output        [boolean] [default: false]
  --local-syntax, -L   Syntax to install packages locally
                                     [string] [default: \\"install --save {list}\\"]
  --global-syntax, -G  Syntax to install packages globally
                                   [string] [default: \\"install --global {list}\\"]
  --help               Show help                                       [boolean]

Examples:
  index.js packages.yaml                    Basic use
  index.js --local=local-packages           Specify target directory
  packages.yaml
  index.js packages.yaml/foo/bar            Install only group foo.bar in
                                            packages.yaml
  index.js /abc/def/packages.yaml/foo/bar   Combine filesystem path and json
                                            path
  index.js file1.yaml/foo/bar1              Filter files and groups
  file1.yaml/foo/bar2 file2.yaml/foo/bar1
  file2.yaml/foo/bar2
  index.js *.yaml/foo/{bar1,bar2}           Using blob pattern (UNIX Shell only)

",
}
`;

exports[`program being invoked with neither arguments nor stdin  1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) Invalid type of tree
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/invalid-schema.yaml --local=virtual-env/target 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) Invalid type of tree
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/invalid-schema.yaml input/invalid-syntax.txt --local=virtual-env/target 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 1, column 9:
    abc: def: ghi
            ^
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/invalid-schema.yaml input/invalid-syntax.txt without options 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 1, column 9:
    abc: def: ghi
            ^
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/invalid-schema.yaml without options 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) Invalid type of tree
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/invalid-syntax.txt --local=virtual-env/target 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 1, column 9:
    abc: def: ghi
            ^
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/invalid-syntax.txt without options 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 1, column 9:
    abc: def: ghi
            ^
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/valid.yaml --local=virtual-env/target 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: 
   [ { path: 'Nested/a/', data: [Array] },
     { path: 'Nested/c/d/', data: [Array] },
     { path: 'Nested/c/e/', data: [Array] },
     { path: 'Flat/', data: [Array] },
     { path: 'DividedFlat/a/', data: [Array] },
     { path: 'DividedFlat/i/', data: [Array] },
     { path: 'SelectiveNested/a/b/', data: [Array] },
     { path: 'SelectiveNested/a/e/', data: [Array] },
     { path: 'SelectiveNested/i/', data: [Array] },
     { path: 'SelectiveNested/i/l/q/r/s/', data: [Array] },
     { path: 'SelectiveNested/i/l/q/u/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml input/invalid-schema.yaml --local=virtual-env/target 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) Invalid type of tree
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/valid.yaml input/invalid-schema.yaml input/invalid-syntax.txt --local=virtual-env/target 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 1, column 9:
    abc: def: ghi
            ^
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/valid.yaml input/invalid-schema.yaml input/invalid-syntax.txt without options 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 1, column 9:
    abc: def: ghi
            ^
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/valid.yaml input/invalid-schema.yaml without options 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) Invalid type of tree
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/valid.yaml input/invalid-syntax.txt --local=virtual-env/target 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 1, column 9:
    abc: def: ghi
            ^
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/valid.yaml input/invalid-syntax.txt without options 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 1, column 9:
    abc: def: ghi
            ^
",
  "stdout": "",
}
`;

exports[`program being invoked with paths input/valid.yaml without options 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: 
   [ { path: 'Nested/a/', data: [Array] },
     { path: 'Nested/c/d/', data: [Array] },
     { path: 'Nested/c/e/', data: [Array] },
     { path: 'Flat/', data: [Array] },
     { path: 'DividedFlat/a/', data: [Array] },
     { path: 'DividedFlat/i/', data: [Array] },
     { path: 'SelectiveNested/a/b/', data: [Array] },
     { path: 'SelectiveNested/a/e/', data: [Array] },
     { path: 'SelectiveNested/i/', data: [Array] },
     { path: 'SelectiveNested/i/l/q/r/s/', data: [Array] },
     { path: 'SelectiveNested/i/l/q/u/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/DividedFlat --local=virtual-env/target 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: [ { path: 'a/', data: [Array] }, { path: 'i/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/DividedFlat without options 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: [ { path: 'a/', data: [Array] }, { path: 'i/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/Flat --local=virtual-env/target 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: [ { path: '', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/Flat without options 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: [ { path: '', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/Flat/i/l --local=virtual-env/target 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: 
   [ { path: 'm/', data: [Array] },
     { path: 'o/', data: [Array] },
     { path: 'q/r/s/', data: [Array] },
     { path: 'q/u/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/Flat/i/l without options 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: 
   [ { path: 'm/', data: [Array] },
     { path: 'o/', data: [Array] },
     { path: 'q/r/s/', data: [Array] },
     { path: 'q/u/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/Nested --local=virtual-env/target 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: 
   [ { path: 'a/', data: [Array] },
     { path: 'c/d/', data: [Array] },
     { path: 'c/e/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/Nested without options 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: 
   [ { path: 'a/', data: [Array] },
     { path: 'c/d/', data: [Array] },
     { path: 'c/e/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/Nested/c --local=virtual-env/target 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: [ { path: 'd/', data: [Array] }, { path: 'e/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/Nested/c without options 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: [ { path: 'd/', data: [Array] }, { path: 'e/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/SelectiveNested --local=virtual-env/target 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: 
   [ { path: 'a/b/', data: [Array] },
     { path: 'a/e/', data: [Array] },
     { path: 'i/', data: [Array] },
     { path: 'i/l/q/r/s/', data: [Array] },
     { path: 'i/l/q/u/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with paths input/valid.yaml/SelectiveNested without options 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: 
   [ { path: 'a/b/', data: [Array] },
     { path: 'a/e/', data: [Array] },
     { path: 'i/', data: [Array] },
     { path: 'i/l/q/r/s/', data: [Array] },
     { path: 'i/l/q/u/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with stdin which contain invalid syntax 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 1, column 9:
    abc: def: ghi
            ^
",
  "stdout": "",
}
`;

exports[`program being invoked with stdin which contain valid syntax and schema 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "{ list: 
   [ { path: 'Nested/a/', data: [Array] },
     { path: 'Nested/c/d/', data: [Array] },
     { path: 'Nested/c/e/', data: [Array] },
     { path: 'Flat/', data: [Array] },
     { path: 'DividedFlat/a/', data: [Array] },
     { path: 'DividedFlat/i/', data: [Array] },
     { path: 'SelectiveNested/a/b/', data: [Array] },
     { path: 'SelectiveNested/a/e/', data: [Array] },
     { path: 'SelectiveNested/i/', data: [Array] },
     { path: 'SelectiveNested/i/l/q/r/s/', data: [Array] },
     { path: 'SelectiveNested/i/l/q/u/', data: [Array] } ],
  pnpm: '/home/khai/programming/pnpm-package-group/integration-test/virtual-env/bin/pnpm' }
",
}
`;

exports[`program being invoked with stdin which contain valid syntax but invalid schema 1`] = `
Object {
  "status": 7,
  "stderr": "[ERROR] (schema) Invalid type of tree
",
  "stdout": "",
}
`;
